### ISSUE TITLE
safari: uninstall procedure

### ISSUE THREAD

**Issue by balmas** [alpheios-project/webextension#145]
We need a better understanding of the uninstall procedure for the Safari extension. As @monzug noted , this is confusing. My experience has been that you need to first close Safari, then eject the app from devices, but I don't know if this is an artifact of the development build. We need to investigate the recommended practice for app extensions and make sure we comply, and then document.

**Comment by monzug** [comment]
So far, I haven't been able to uninstall the alpheios app. these are the steps that I followed. from Extensions tab in Safari Preferences, I click on Uninstall the reading tools. I get the message that I have to remove the Alpheios app. I close safari and quit Alpheios app (no longer available from the dock). then I try to remove the Alpheios app from the application folder but I get: alpheios app can't be moved to the trash because it can't be deleted. then I try to remove the dmg file but I get an other message that says that if I want to delete from my computer I have to use selective sync. This is where I stopped.

**Comment by irina060981** [comment]
I have spent some time to find out how the uninstall procedure of safari app extension should be performed now.
And it was changed since Safari extensions became Safari App Extension.

Now we should remove them manually.

So There are the following steps (if it was installed from dmg):

1) close AlpheiosApp
2) close Safari
3) eject AlpheiosApp from devices

If it was installed from XCode debug process - simply remove files from Debug folder (it would be opened using Uninstall button).

The button Uninstall could only show where the Safari App Extension is placed. But first we should close all applications that using it.
Here it is a discussion on the forum
https://forums.developer.apple.com/thread/76486

**Comment by monzug** [comment]
Hi Irina, I come up with the same conclusion (I guess based on same investigation). so, here are few questions:
1) if the reader cannot be uninstall, the uninstall button shouldn't be there. correct?
2) as said above, I follow the steps suggested but I could not remove the alpheios app manually. how do I remove the app then?

**Comment by irina060981** [comment]
Hello, Monica!

1. As Apple staff said in the forum - 
"There should be an Uninstall button for Safari App Extensions in the latest beta of macOS High Sierra 10.13.4. It won't move your app to the Trash automatically, but it will reveal it in the Finder and tell the user how to perform the action." 
So the button Uninstall exists, but it doesn't uninstall.

2. I was able to remove manually after - 
    - I have closed Alpheios Application (be sure that you have closed it using Quit (command + Q) and it is not in the bottom panel
    - I have closed Safari (close button is not enough - you should Quit Safari (command+Q) and be sure it doesn't active in the bottom panel.
    I have spent some time to understand that close not equal quite :)

I am a newbie with MacOS and Safari - so I don't know if it is a logical behaviour for Mac or not :) So I didn't mention those details - I am sorry :)
Hope it will help!

**Comment by monzug** [comment]
hahaha, we are all discovering a new world!
ok, I am going to try with Safari quit and not close. THX

**Comment by monzug** [comment]
no luck. Quit Safari and Alpheiosapp but could not uninstall it. why don't
we have the alpheiosapp icon listed in the Application folder?

**Comment by balmas** [comment]
I think we won't get that until we actually have an unrestricted app to install. We aren't yet building with a production profile. They don't make it to easy to test this install process.

@irina060981 I found some instructions at https://help.apple.com/xcode/mac/current/#/dev16fb2b911 that might be helpful here.

**Comment by irina060981** [comment]
@monzug , I am upset that there is no luck with it. But I couldn't understand - what exactly the error do you get trying to eject .dmg with AlpheiosApp?
I have re-read you first steps descriptions and you pointed something like the following:
" then I try to remove the dmg file but I get an other message that says that if I want to delete from my computer I have to use selective sync"
Could you make a screen with the error? May be the error is not from our development process?
I don't know what is selective sync and couldn't find something appropriate with Google.

@balmas , I tried to follow the steps from the instruction - but I have no appropriate certificates:
<img width="805" alt="screen shot 2018-11-07 at 09 49 45" src="https://user-images.githubusercontent.com/12377640/48101621-6184a700-e273-11e8-92d6-38b3b4a64e91.png">

**Comment by monzug** [comment]
here we are
![screen shot 2018-11-02 at 1 14 41 pm](https://user-images.githubusercontent.com/41396793/48116818-4135f780-e235-11e8-8121-d79da2099611.png)
![screen shot 2018-11-02 at 1 15 57 pm](https://user-images.githubusercontent.com/41396793/48116820-4135f780-e235-11e8-87b0-e3cdb0f42cf4.png)

**Comment by irina060981** [comment]
Hello, Monica!

It seems to me that you are trying to delete something not what should be deleted for uninstall, or may be you have some another installation process.

How I do this:

1) I could use built files from AppSafari folder (not .dmg file) from my build
<img width="273" alt="screen shot 2018-11-07 at 21 35 36" src="https://user-images.githubusercontent.com/12377640/48129403-18653f00-e2d5-11e8-8b58-95cdc044bac6.png">

So to add Safari App Extension to Safari - I simply should launch AlpheiosApp from the folder
If I will click on Uninstall button for the extension in Extension's tab - it will open the same folder - AlpheiosApp - with the same files.
<img width="862" alt="screen shot 2018-11-07 at 21 37 43" src="https://user-images.githubusercontent.com/12377640/48129485-62e6bb80-e2d5-11e8-87cc-9a7eb74f15a8.png">

If I will try to delete them - I won't be able - because they are currently in use.
But if I will close the application - I would be able to delete all 4 files from folder and extension will disappear from Extension's tab.

2) I could install from .dmg and then uninstall process will be a little other.

**Comment by irina060981** [comment]
No, I am creating builds on my Mac and install them from my local machine.
May be there is some conflict with deleting/ejecting data from network.
I will try to do it tomorrow from some dropbox account to get the same errors.

**Comment by monzug** [comment]
I'll send a screenshot that will give you more inside. will talk tomorrow.
will be available earlier in the day.

**Comment by balmas** [comment]
@irina060981 i sent your email an invite to the Dropbox folder that i have been using the share builds with Monica.

**Comment by irina060981** [comment]
Sorry, @balmas and @monzug  , I have missed it in email box! I will test with it today!

**Comment by irina060981** [comment]
I have no dropbox application and I haven't found any free variant.
So I downloaded last .dmg (and it said me that it is from unidentified developer), changed security properties and installed.
And then uninstalled the same as in 2) workflow - it was removed without any problems.

**Comment by monzug** [comment]
So, I delete my share folder with the alpheios app and then somehow I was able to remove the app. there are still files there so I might need to delete them one by one. so it's gone but not in a 'clean' way...

**Comment by balmas** [comment]
Ok, so we should probably not waste too much time on issues that are specific to how I was sharing the build since this isn't how the actual release will be distributed...

@monzug maybe it would be better to copy the dmg file out of Dropbox onto your home directory or something on your mac, before installing it?

@irina060981 I was using Dropbox just because I didn't want to put binaries into GitHub, but we need a solution for sharing the builds that works for all of us. Do you mean you cannot access the web-based Dropbox.com with a free account?

**Comment by irina060981** [comment]
@balmas , I could download it locally.
May be I don't know how to use from share folder?

**Comment by balmas** [comment]
That's interesting. My mac came with a Dropbox client already installed.  In any event, I don't think it's worth worrying about this too much. When I get back I will finish creating the production certificate and profile so that we can try to get closer to the user experience of installation.

**Comment by irina060981** [comment]
oh, I have found how to install dropbox from the web-site.
I have installed it - so I have dropBox attached to the MAC

Then I have installed dmg from it, then launched application, turned on extension in Safari.
Then closed the application, closed Safari (quited of course :) ) and then ejected.
No errors

It seems to me that it is something specific with some sync procedure on Monica's MAC.

**Comment by monzug** [comment]
agree. it's most likely my environment. @balmas, I did install both way,
locally and from dropbox. I was trying to replicate Harry's issues with no
luck....

**Comment by monzug** [comment]
Irina,
after installing the latest safari build, did you experience any delays in
starting the alpheiosapp? Thx

**Comment by irina060981** [comment]
Monica, there was a bug with exchanging messages between content and background.
@kirlat has fixed it today.
May be those delays were because of that bug.
I have seen the delay - but as I am switching between several projects I thought that it was always and I simply forgot about it.

I could try to create a new build and upoad to dropbox - I think I use the correct certificate but I haven't checked it :)
Do you need a new build?

**Comment by monzug** [comment]
let's wait for @balmas to get back and get a new build. then I will check
again if there is still the delay. Thanks

**Comment by irina060981** [comment]
I don't see a delay in a new build (I was testing it myself locally) - it opens and closes much faster.
But my eyes are not very objective :)

**Comment by monzug** [comment]
good to know! will wait the new build then. Thx

**Comment by irina060981** [comment]
ok - let's wait for Bridget! Hope she has a great time in Belgium! :)

**Comment by balmas** [comment]
ok, assigning this one back to Monica then to test with build 2.1.0.7 and then we'll see where we are.

**Comment by monzug** [comment]
well, today I spent quite sometimes to install the 2.1.0.7. eventually it worked but honestly I do not know why. I did repeat the same steps 10 times and then it worked... I keep this one open as we want to retest this when we will have an installation (and uninstall) closer to user experience.

**Comment by irina060981** [comment]
Thank you, Bridget, for sharing this information :) So much details for working with Apple that could "messes up Apple's own engineers"

**Comment by balmas** [comment]
a few more notes on the install process:

1) I have run the validation process in XCode (https://help.apple.com/xcode/mac/current/#/dev37441e273) which is supposed to validate that the app meets minimum requirements and passes standard app store connect tests

2) I have also run the package installation test as described at https://help.apple.com/xcode/mac/current/#/dev16fb2b911 and that passed as well, putting the app into the /Applications folder.  However I could not actually run the installed app because it crashed with a signing error. I think this might be expected, since it hasn't been signed by Apple as production ready, but I have no idea. it's impossible to tell from the documentation.

3) It seems that there is no way to test the actual user experience of installing the app extension from the app store ahead of time. They make a test flight available for iOS apps but not Mac. I have sent an email to their support asking if there is any way to test the actual user experience ahead of time. Am not optimistic of the answer.

4) The only other option I see is to try testing a manually distributed notarized package which isn't restricted by device id as described at https://help.apple.com/xcode/mac/current/#/dev033e997ca. This seems possible , but requires a bit of work, and in the end isn't really any closer to testing the app store install process than our current method of the developer distribution. So I'm not really enthusiastic about putting time into it. 

At this point, I am inclined to just put our faith in the app store distribution process to detect errors with installation and uninstallation and focus our testing on the app functionality itself, once installed. It seems fairly likely that they will be happy to fail our app in the review process if it doesn't meet their standards. 

@monzug what do you think?

**Comment by monzug** [comment]
How much you like the Apple world????? I am wondering how the other apps
do. I trust you,@balmas

**Comment by monzug** [comment]
as you mentioned: TestFlight supports apps for iOS, watchOS, tvOS, and
iMessage.

**Comment by irina060981** [comment]
I hope that Apple team will be interested in to help us in developing a Safari App Extension as there are few of them now - because of there ignoring standards and making the development workflow so immature.
So they should be interested to popularize it while there tools are still in alpha/beta modes.
Oh how are they going to re-developing Safari App Extensions in there App Store after removing all legacy extensions supoort?
I hope :)

**Comment by monzug** [comment]
sure

**Comment by balmas** [comment]
an installer package for build 2.1.0.16 is available in the Dropbox folder, named Alpheios.pkg

To test this install you need to 
1. Make sure any development build of Alpheios are ejected and removed from your home directory and gone from Safari.
2. Run the following command in a terminal: `sudo installer -store -pkg Alpheios.pkg -target /`

What you should see is
- [ ] that the Alpheios app is now installed in your Applications folder
- [ ] You should be able to run it, see that it adds Alpheios to the Safari Extensions (not checked by default)
- [ ] You should be able to enable the Alpheios extension and run it as normal
- [ ] You should also be able to uinstall the Alpheios extension and remove the app from the Applications folder

**Comment by monzug** [comment]
first, the last / backslash should be removed
then I am stuck here

![screen shot 2018-12-12 at 8 51 36 am](https://user-images.githubusercontent.com/41396793/49854727-7234b980-fdeb-11e8-9414-4eed86bf1a02.png)

**Comment by monzug** [comment]
as per installation process, Alpheios is checked in the Safari Extensions

**Comment by balmas** [comment]
ok, I'm going to consider this done then. Thanks!

**Comment by monzug** [comment]
how about the Alpheios extension checked at installation. I had to uncheck
and check again to run it.

**Comment by balmas** [comment]
I am not going to worry too much about this. I suspect it it is some artifact of the fact that we are testing with different versions of the extension installed in different ways. As we cannot really test what the user experience is here, I think we will have to rely on the Apple review and validation mechanisms to make sure our install package works as it should.

**Comment by monzug** [comment]
ok. let me do an other installation before giving a go.
